segmentation/utils/synthetic_spatial_datagen.py

```diff
@@ -136,37 +134,91 @@ def point_init_dist_matrix(size_img=(1024, 1024), num_A=100, num_B=100, num_C=10
+    than those of type C
+
+    We will use a multivariate Gaussian distribution for A, B, and C type cells to generate their respective centers.
 
-    # an astronomically unlikely error, would occur if for some reason all of the points were labelled as duplicate
-    if len(non_dup_points) == 0:
-        raise ValueError("Bad run: no unique points generated. Try again, will work next time.")
+    Args:
+        size_img: a tuple indicating the size of the image. Default 1024 x 1024
+        num_A: the number of A centroids to generate. Default 100.
+        num_B: the number of B centroids to generate. Default 100.
+        num_C: the number of C centroids to generate. Default 100.
 
-    point_x_coords, point_y_coords = zip(*non_dup_points)
+        mean_A_factor: a tuple to determine which number to multiply the height and width by
+            to indicate the center (mean) of the distribution to generate A points.
+            Will be randomly set to a predefined value if None.
+        cov_A: the covariance used to generate A poins in the format [[varXX, varXY], [varYX, varYY]].
+            Will be randomly set to a predefined value if None.
+        mean_B_factor: similar to mean_A_factor
+        cov_B: similar to cov_A
+        mean_C_factor: similar to mean_A_factor
+        cov_C: similar to cov_A
 
-    # generate the binary matrix to pass into label_map
-    binary_mat = np.zeros(size_img)
-    binary_mat[point_x_coords, point_y_coords] = True
+        seed: whether to fix the random seed or not. Useful for testing.
+            Should be a specified integer value. Default None.
 
-    # generate the label matrix for the image now
-    label_mat = label(binary_mat)
+    Returns:
+        sample_img_xr: the data in xarray format containing the randomized label matrix
+            based on the randomized centroid centers we generated. The label mat portion
+            of sample_img_xr is generated from a randomly initialized set of cell centroids
+            where those of type a are on average closer to those of type b than they
+            are to those of type c.
+    """
 
-    # and create the output to be able to run through calc_dist_matrix
-    # for now, I'm assuming that the array returned will have just one fov
-    # and we don't know anything about the segmentation labels
+    # generate the list of centroids and zip them into x and y coords
+    all_centroids = \
+        generate_random_centroids(size_img=size_img, num_A=num_A, num_B=num_B, num_C=num_C,
+                                  mean_A_factor=mean_A_factor, cov_A=cov_A,
+                                  mean_B_factor=mean_B_factor, cov_B=cov_B,
+                                  mean_C_factor=mean_C_factor, cov_C=cov_C,
+                                  seed=seed)
+
+    point_x_coords, point_y_coords = zip(*all_centroids)
+
+    # to compute the indices of the original centroid locations, needed to adjust the distance matrix
```

Comment:
I'm not sure I understand this comment

Reply:
Ah yes, need to make this clearer.

Basically, on the call to regionprops in calc_dist_matrix, the function automatically orders the centroids in ascending order regardless of the original label assigned to them by the label function (or whatever method you use to label the centroids). So for example, if you labeled the centroid (1, 1) as number 75 in the corresponding label map for dist_mat_pos, regionprops would automatically put that as the 0th element instead. Do this for all centroids and you've got a really jumbled distance matrix that would not reflect the first m columns being marker 1 cells, the next n columns being marker 2 cells, and the final p columns being random marker cells.

I didn't want to mess around with Jay's calc_dist_matrix function too much, so the easy fix is to compute the original locations (aka indices) of where the centroids should go (the order in which they appear in all_centroids of generate_test_label_map, I intentionally have first the a, then the b, lastly the c centroids). And that's the beauty of lexsort, knowing that regionprops sorts the centroids in ascending order, first by x-coord, then y-coord. lexsort will essentially return the indices needed to reorder a numpy array altered by np.sort back to what it was originally. That way, we can reorder the rows and columns of the distance matrix to accurately reflect the partitioning we desire.